src/feature/hs/hs_dos.c

```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2019, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file hs_dos.c
+ **/
+
+#define HS_DOS_PRIVATE
+
+#include "core/or/circuitlist.h"
+
+#include "hs_dos.h"
+
+/*
+ * Public API.
+ */
+
+/* Return true iff an INTRODUCE2 cell can be sent on the given service
+ * introduction circuit. */
+bool
+hs_dos_can_send_intro2(or_circuit_t *s_intro_circ)
+{
+  tor_assert(s_intro_circ);
+  tor_assert(TO_CIRCUIT(s_intro_circ)->purpose ==
+             CIRCUIT_PURPOSE_INTRO_POINT);
+
+  /* This is called just after we got a valid and parsed INTRODUCE1 cell. The
+   * service has been found and we have its introduction circuit.
+   *
+   * First, the INTRODUCE2 bucket will be refilled (if any). Then, decremented
+   * because we are about to send or not the cell we just got. Finally,
+   * evaluate if we can send it based on our token bucket state. */
+
+  /* Refill INTRODUCE2 bucket. */
+  token_bucket_ctr_refill(&s_intro_circ->introduce2_bucket, approx_time());
+
+  /* Decrement the bucket for this valid INTRODUCE1 cell we just got. Don't
+   * underflow else we end up with a too big of a bucket. */
+  if (token_bucket_ctr_get(&s_intro_circ->introduce2_bucket) > 0) {
```

Comment:
That's another comment towards the token bucket API, but it's a bit weird that we have to explicitly check fort underflow like this. Is there any case where we would want the token bucket to underflow? If not, can't we handle this inside the decrement method?

Reply:
Yeah so I think a series of improvement could go towards the token bucket API. The DoS subsystem (dos.c) does way more to prevent over/underflows and handle clock jumps.

Reply:
Wanna open a  token bucket API ticket for this and the above comments? If we get a ticket, it's fine to not fix this right here.

Reply:
Yes definitely. I've noted it down. Once merged, we need more robust token bucket checks.